internal/format/transformer.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package format
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/emicklei/proto"
+	"github.com/uber/prototool/internal/settings"
+	"github.com/uber/prototool/internal/text"
+	"go.uber.org/zap"
+)
+
+type transformer struct {
+	logger *zap.Logger
+}
+
+func newTransformer(options ...TransformerOption) *transformer {
+	transformer := &transformer{
+		logger: zap.NewNop(),
+	}
+	for _, option := range options {
+		option(transformer)
+	}
+	return transformer
+}
+
+func (t *transformer) Transform(config settings.Config, data []byte) ([]byte, []*text.Failure, error) {
+	descriptor, err := proto.NewParser(bytes.NewReader(data)).Parse()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// log statements are at debug level so
+	// this will trigger if debug is set
+	// TODO maybe remove
+	logVisitor := newLogVisitor(t.logger)
+	for _, element := range descriptor.Elements {
+		element.Accept(logVisitor)
+	}
+
+	firstPassVisitor := newFirstPassVisitor(config)
+	for _, element := range descriptor.Elements {
+		element.Accept(firstPassVisitor)
+	}
+	failures := firstPassVisitor.Do()
+	buffer := bytes.NewBuffer(nil)
```

Comment:
While this will technically work here, the returned `Bytes()` value is only valid until the next modification in `firstPassVisitor` - while we don't do another modification, I don't like relying on these semantics and prefer to make a copy here.

Reply:
yeah good call, just read the `NewBuffer` doc